Configure OpenAI models in the app host

sebastienros opened the pull request:
Building over [@SteveSandersonMS PR](https://github.com/dotnet/aspire/pull/6225) to demonstrate the end-to-end experience when ai models/deployments are configured at the host level only.
 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/6577)

SteveSandersonMS committed:
Add AddAzureOpenAIChatClient

SteveSandersonMS committed:
Add AddAzureOpenAIChatClient

SteveSandersonMS committed:
Set exact dependency versions in MEAI

SteveSandersonMS committed:
Set exact dependency versions in MEAI

SteveSandersonMS committed:
Fix typo

SteveSandersonMS committed:
Fix typo

SteveSandersonMS committed:
Handle connection string missing values

SteveSandersonMS committed:
Handle connection string missing values

SteveSandersonMS committed:
Handle connection string including duplicate values

SteveSandersonMS committed:
Handle connection string including duplicate values

SteveSandersonMS committed:
Change to chaining API

SteveSandersonMS committed:
Change to chaining API

sebastienros committed:
Merge branch 'main' into stevesa/meai-integration

sebastienros commented on src/Aspire.Hosting.Azure.CognitiveServices/AzureOpenAIExtensions.cs:
```diff
@@ -103,14 +106,74 @@ public static IResourceBuilder<AzureOpenAIResource> AddAzureOpenAI(this IDistrib
         return builder;
     }
+
+    /// <summary>
```
Copy-paste, ignore

sebastienros commented on src/Aspire.Hosting.Azure.CognitiveServices/AzureOpenAIConnectionResource.cs:
```diff
@@ -0,0 +1,36 @@
+/// <summary>
+/// Represents an Azure OpenAI resource.
+/// </summary>
+public class AzureOpenAIConnectionResource(ParameterResource innerResource, string? environmentVariableName) : IHasDeploymentResource
```
Resource to represent an open ai resource as a connection string only.

eerhardt replied:
Why do we need this? Why can't we just use the AzureOpenAIResource?

sebastienros replied:
I wanted a way to configure an existing resource without provisioning. Similar to https://github.com/dotnet/aspire/issues/2395.

sebastienros replied:
Removed, out of scope

eerhardt commented on src/Aspire.Hosting.Azure.CognitiveServices/PublicAPI.Shipped.txt:
```diff
@@ -14,4 +14,3 @@ Aspire.Hosting.ApplicationModel.AzureOpenAIResource.Deployments.get -> System.Co
 Aspire.Hosting.AzureOpenAIExtensions
 static Aspire.Hosting.AzureOpenAIExtensions.AddAzureOpenAI(this Aspire.Hosting.IDistributedApplicationBuilder! builder, string! name) -> Aspire.Hosting.ApplicationModel.IResourceBuilder<Aspire.Hosting.ApplicationModel.AzureOpenAIResource!>!
 static Aspire.Hosting.AzureOpenAIExtensions.AddAzureOpenAI(this Aspire.Hosting.IDistributedApplicationBuilder! builder, string! name, System.Action<Aspire.Hosting.ApplicationModel.IResourceBuilder<Aspire.Hosting.ApplicationModel.AzureOpenAIResource!>!, Aspire.Hosting.ResourceModuleConstruct!, Azure.Provisioning.CognitiveServices.CognitiveServicesAccount!, System.Collections.Generic.IEnumerable<Az ...
-static Aspire.Hosting.AzureOpenAIExtensions.AddDeployment(this Aspire.Hosting.ApplicationModel.IResourceBuilder<Aspire.Hosting.ApplicationModel.AzureOpenAIResource!>! builder, Aspire.Hosting.ApplicationModel.AzureOpenAIDeployment! deployment) -> Aspire.Hosting.ApplicationModel.IResourceBuilder<Aspire.Hosting.ApplicationModel.AzureOpenAIResource!>!
```
Are we able to do this without making breaking changes? For example, can we obsolete something and add new APIs?

sebastienros replied:
I reverted as I didn't remember why the change was done. Do we want to keep `AddDeployment` or replaced it with `WithDeployment`?

SteveSandersonMS commented on playground/OpenAIEndToEnd/OpenAIEndToEnd.WebStory/Components/Pages/Home.razor:
```diff
@@ -1,39 +1,36 @@
-    private ChatClient? chatClient = null;
-    private List<ChatMessage> chatMessages = new List<ChatMessage>();
+    [Inject(Key = "modelA1")]
+    public required IChatClient chatClient { get; set; }
```
I think in this page we should only use `ChatClient` and not `IChatClient`, because it's meant to demonstrate usage of the concrete OpenAI client types directly rather than going through MEAI abstractions. It's the other page, `UseIChatClient`, that's meant to show usage of MEAI abstractions.

sebastienros replied:
Reverted the changes. Now testing with the `/useichatclient` page you added

SteveSandersonMS commented on src/Aspire.Hosting.Azure.CognitiveServices/IHasDeploymentsResource.cs:
```diff
@@ -0,0 +1,20 @@
+/// <summary>
+/// Represents an Azure OpenAI resource.
+/// </summary>
+public interface IHasDeploymentResource: IResourceWithConnectionString, IResourceWithEnvironment
```
This is a surprisingly general type name - `Aspire.Hosting.ApplicationModel.IHasDeploymentResource` doesn't sound specific to Azure OpenAI but from the implementation it is. Could this be named something more specific to AzureOpenAI?

SteveSandersonMS commented on src/Components/Aspire.Azure.AI.OpenAI/AzureOpenAISettings.cs:
```diff
@@ -70,6 +70,11 @@ public bool DisableTracing
     }
 
+    /// <summary>
+    /// Gets the models to use for the Azure OpenAI resource.
```
Can the comment explain what the keys and values represent? It's not clear from the property signature alone.

SteveSandersonMS commented:
Overall looks great. @eerhardt's point about the breaking change is probably the main thing to consider. Besides that it looks pretty close to being ready.

sebastienros committed:
Configure OpenAI models in the app host

sebastienros committed:
Merge remote-tracking branch 'origin/stevesa/meai-integration' into sebros/hostmodels

# Conflicts:
#	playground/OpenAIEndToEnd/OpenAIEndToEnd.WebStory/Program.cs
#	src/Components/Aspire.Azure.AI.OpenAI/AspireAzureOpenAIClientBuilderChatClientExtensions.cs
#	src/Components/Aspire.Azure.AI.OpenAI/PublicAPI.Unshipped.txt

sebastienros committed:
Restore custom configuration

sebastienros commented:
I updated the PR. @SteveSandersonMS is this something we could merge in your PR as it's impacting how configuration is passed to the clients. Other option is I wait for your PR to be done and then react to the changes.

SteveSandersonMS commented:
@sebastienros I'm fine with it being merged into my PR, but my PR itself can't be completed until we get the next build of Microsoft.Extensions.AI, because the agreed API shape depends on API changes we've just merged into `dotnet/extensions` today.

sebastienros commented:
@eerhardt does it make sense to split this PR in two: the deployment API changes creating the ENVs, and the MEAI part that is using it and would be part of Steve's.

Or just move everything to Steve's branch and ship everything at once.

eerhardt commented:
> @eerhardt does it make sense to split this PR in two: the deployment API changes creating the ENVs, and the MEAI part that is using it and would be part of Steve's.
> 
> Or just move everything to Steve's branch and ship everything at once.

Completely up to you and Steve. I don't see a reason why they would need to be separate, but if you think they should be, I wouldn't argue.

danmoseley commented:
@SteveSandersonMS is this PR still relevant/in progress?

SteveSandersonMS commented:
@danmoseley It's @sebastienros's PR, so he should comment, but I think the functionality could still be relevant. I don't think it's urgent or a pre-req for anything else we have going on though.

eerhardt commented:
@dotnet-policy-service - please keep this PR open.

eerhardt committed:
Merge remote-tracking branch 'upstream/main' into sebros/hostmodels

eerhardt committed:
Add OpenAI child deployment resource.

Use it in the OpenAI playground.

eerhardt commented on src/Aspire.Hosting.Azure.CognitiveServices/AzureOpenAIExtensions.cs:
```diff
@@ -14,6 +14,8 @@ namespace Aspire.Hosting;
 /// </summary>
 public static class AzureOpenAIExtensions
 {
+    internal const string DefaultConfigSectionName = "Aspire:Azure:AI:OpenAI";
+
```
```suggestion
```

sebastienros commented on src/Aspire.Hosting.Azure.CognitiveServices/AzureOpenAIExtensions.cs:
```diff
@@ -116,8 +100,44 @@ public static IResourceBuilder<AzureOpenAIResource> AddAzureOpenAI(this IDistrib
     /// <summary>
-    /// Adds an Azure OpenAI Deployment resource to the application model. This resource requires an <see cref="AzureOpenAIResource"/> to be added to the application model.
+    /// Adds an Azure OpenAI Deployment to the <see cref="AzureOpenAIResource"/> resource. This resource requires an <see cref="AzureOpenAIResource"/> to be added to the application model.
     /// </summary>
```
Can we obsolete this one now or does it still server a purpose?

eerhardt replied:
That is a good topic for discussion. On one hand, we don't necessarily _have to_ obsolete it. But telling people "hey you should move from this old API to new AI" seems useful. Thoughts?

eerhardt committed:
PR feedback

- Obsolete the existing APIs
- Add unit tests
- Switch the deployment name to default to the resource name

eerhardt committed:
Revert unnecessary changes

eerhardt committed:
PR feedback

Change CreateDeployment to not take a ref and return the same thing.
